### PR TITLE
[HZ-907] Protect MulticastDiscoveryStrategy against Java deserialization attacks

### PR DESCRIPTION
**Pull request by kwart**
This PR adds an alternative and safer serialization format for `MulticastDiscoveryStrategy` datagrams.

By using the property configuration `safe-serialization=true`, users might protect themselves against Java deserialization-based attacks.

### CODE REVIEW

**Review by vbekiaris — COMMENTED** [review]
The PR looks good. The new safe-serialization format is not backwards compatible (context: EE rolling upgrades), can we do better in this area?

File: hazelcast/src/test/java/com/hazelcast/spi/discovery/multicast/MulticastDiscoveryStrategyDeserializationTest.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    private Config createConfig(Boolean safeSerialization) {
+        Config config = smallInstanceConfig().setProperty("hazelcast.discovery.enabled", "true");
+        DiscoveryStrategyConfig dsc = new DiscoveryStrategyConfig(MulticastDiscoveryStrategy.class.getName())
+                .addProperty("group", GROUP).addProperty("port", String.valueOf(PORT));
+        if (safeSerialization != null) {
+            dsc.addProperty("safe-serialization", safeSerialization.toString());
+        }
+        JoinConfig joinConfig = config.getNetworkConfig().getJoin();
+        joinConfig.getDiscoveryConfig().addDiscoveryStrategyConfig(dsc);
+        // joinConfig.getAutoDetectionConfig().setEnabled(false);
```
Comment (olukas) [inline hazelcast/src/test/java/com/hazelcast/spi/discovery/multicast/MulticastDiscoveryStrategyDeserializationTest.java:114]: leftover?

**Review by olukas — APPROVED** [review]
LGTM, just one very minor comment.

File: hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/discovery/ClientToMemberDiscoveryTest.java [hunk 2]
```diff
@@ -64,8 +76,14 @@ public void tearDown() {
...
-        instance1 = factory.newHazelcastInstance(serverConfig);
-        instance2 = factory.newHazelcastInstance(serverConfig);
+        if (safeSerialization != null) {
+            serverConfig.getNetworkConfig().getJoin().getDiscoveryConfig().getDiscoveryStrategyConfigs().iterator().next()
+                    .addProperty("safe-serialization", safeSerialization.toString());
+            clientConfig.getNetworkConfig().getDiscoveryConfig().getDiscoveryStrategyConfigs().iterator().next()
+                    .addProperty("safe-serialization", safeSerialization.toString());
+        }
+        factory.newHazelcastInstance(serverConfig);
+        factory.newHazelcastInstance(serverConfig);
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/discovery/ClientToMemberDiscoveryTest.java:86]: What do the error messages look like when there is a configuration mismatch between clients and servers? (it is enabled on the server, but the user forgets to enable it on the client)

**Review by mdumandag — APPROVED** [review]
LGTM with a question

**Review by vbekiaris — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/spi/discovery/multicast/MulticastProperties.java [hunk 3]
```diff
@@ -39,6 +40,22 @@
...
+     *
+     * <pre>
+     * boolean: flag if memberInfo provided (false mean memberInfo is null)
+     * UTF-8:   host
+     * int:     port
+     * </pre>
+     *
+     * @since 5.1
+     */
+    public static final PropertyDefinition SAFE_SERIALIZATION = property("safe-serialization", BOOLEAN);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/discovery/multicast/MulticastProperties.java:57]: can we have a concise description of what we mean by safe-serialization here as javaDoc? Better if we include an example breach scenario in it.

File: hazelcast/src/test/java/com/hazelcast/spi/discovery/multicast/MulticastDiscoveryStrategyDeserializationTest.java (hunk 1 above)
Comment (kwart) [reply hazelcast/src/test/java/com/hazelcast/spi/discovery/multicast/MulticastDiscoveryStrategyDeserializationTest.java:114]: :+1:

File: hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/discovery/ClientToMemberDiscoveryTest.java (hunk 2 above)
Comment (kwart) [reply hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/discovery/ClientToMemberDiscoveryTest.java:86]: Users without the `safe-serialization` will see the warning message in the log on both sides: client and member.

```
The MulticastDiscoveryStrategy Hazelcast member discovery strategy is configured without the safe-serialization parameter enabled. Set the safe-serialization property to 'true' in the strategy configuration to protect the cluster against untrusted deserialization attacks.
```

If the `safe-serialization` value doesn't match on client and member, the discovery fails:
```
09:49:55,412  INFO |clientSerializationMismatchTest[1: safeSerialization: false]| - [ClientConnectionManager] clientSerializationMismatchTest - hz.client_3 [dev] [5.1-SNAPSHOT] Trying to connect to cluster: dev
09:49:57,267  WARN |clientSerializationMismatchTest[1: safeSerialization: false]| - [ClientConnectionManager] clientSerializationMismatchTest - hz.client_3 [dev] [5.1-SNAPSHOT] Unable to get live cluster connection, retry in 1000 ms, attempt: 1, cluster connect timeout: INFINITE, max backoff: 30000 ms
...
```

File: hazelcast/src/main/java/com/hazelcast/spi/discovery/multicast/MulticastProperties.java (hunk 3 above)
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/spi/discovery/multicast/MulticastProperties.java:57]: I will extend the description and add a link to an OWASP page related to the deserialization.
I don't want to share exactly the attack vector as we don't do that for found vulnerabilities.

**Review by ahmetmircik — APPROVED** [review]

**Pull request merged**